### PR TITLE
Search: support datasource template variables when parsing dashboard JSON models

### PR DESCRIPTION
**Pull request by ArturWierzbicki**
This PR fixes parsing the Dashboard JSON DB models into SearchV2's `DashboardInfo` DTOs when dashboards use datasource template variables

### CODE REVIEW

**Comment by ArturWierzbicki** [comment]
There is one more case to handle with `$__all`: 

image

image


In this case, the actual UIDs are nowhere to be found in the JSON model so I am not exactly sure what to do - filter all matching datasources from DB?

File: pkg/services/searchV2/extract/dashboard.go [hunk 1]
```diff
@@ -139,6 +227,35 @@ func ReadDashboard(stream io.Reader, lookup DatasourceLookup) (*DashboardInfo, e
 		}
 	}
 
+	for i, panel := range dash.Panels {
```
Comment (ryantxu) [inline pkg/services/searchV2/extract/dashboard.go:230]: Can you add a comment explaining what this does?  I *think* something like:
```
replace variable datasource references with concrete values
```

File: pkg/services/searchV2/extract/dashboard.go [hunk 2]
```diff
@@ -139,6 +227,35 @@ func ReadDashboard(stream io.Reader, lookup DatasourceLookup) (*DashboardInfo, e
...
+				variableName = strings.TrimPrefix(strings.TrimSuffix(variableName, "}"), "${")
+			} else {
+				variableName = strings.TrimPrefix(variableName, "$")
+			}
+
+			refs := datasourceVariablesLookup.getDatasourceRefs(variableName)
+			referencedDs = append(referencedDs, refs...)
+		}
+
+		dash.Panels[i].Datasource = append(dsRefs, referencedDs...)
```
Comment (ryantxu) [inline pkg/services/searchV2/extract/dashboard.go:256]: Is this the same thing?
```suggestion
		for _, dsVariableRef := range dsVariableRefs {
			variableName := dsVariableRef.UID
			if strings.HasPrefix(variableName, "${") {
				variableName = strings.TrimPrefix(strings.TrimSuffix(variableName, "}"), "${")
			} else {
				variableName = strings.TrimPrefix(variableName, "$")
			}

			refs := datasourceVariablesLookup.getDatasourceRefs(variableName)
			dsRefs = append(dsRefs, refs...)
		}

		dash.Panels[i].Datasource = dsRefs
```

File: pkg/services/searchV2/extract/dashboard.go (hunk 1 above)
Comment (ArturWierzbicki) [reply pkg/services/searchV2/extract/dashboard.go:242]: extracted some fns

File: pkg/services/searchV2/extract/dashboard.go (hunk 2 above)
Comment (ArturWierzbicki) [reply pkg/services/searchV2/extract/dashboard.go]: I'd hope so, but right now the `Panel` model `Datasource` prop is an array, so I guess some of them could be variables and some of them could be proper references. I can't think of a scenario in which thats possible though

image

**Review by ryantxu — APPROVED** [review]
nice -- I pushed updates for the export section.

**Pull request merged**